### ISSUE TITLE
No Confusion matrix generated for both normalized and other one when i was trying YOLOv8 OBB

### ISSUE THREAD

**Issue by FahadRamzan** [ultralytics/ultralytics#8256]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

No Confusion matrix generated for both normalized and other one when i was trying YOLOv8 oBb? 

![confusion_matrix_normalized](https://github.com/ultralytics/ultralytics/assets/88324452/4aed4450-f38a-4cb4-ab4b-08b851683274)
![confusion_matrix](https://github.com/ultralytics/ultralytics/assets/88324452/0eb52361-d8af-466c-a51e-5d572723e7cb)

There are other graphs that are generated such as curves for f1 precision recall

**Comment by FahadRamzan** [comment]
Can we print out these metric like this? Is there any way possible?

precision = TP / (TP + FP)
recall = TP / (TP + FN)
f1_score = 2 * (precision * recall) / (precision + recall)
accuracy = (TP + TN) / (TP + FP + FN + TN)

**Comment by glenn-jocher** [comment]
@FahadRamzan the confusion matrix is a valuable tool for understanding the performance of a classification model, including object detection models like YOLOv8. It provides insights into the types of errors made by the model, such as which classes are being confused with one another.

If the confusion matrix is not being generated during your validation process with YOLOv8 OBB (Oriented Bounding Box), there could be several reasons for this issue:

1. **Validation Settings**: Ensure that the validation settings are correctly configured to generate the confusion matrix. You might need to set specific flags or parameters to enable the creation of the confusion matrix.

2. **Data Issues**: Check if the validation dataset is correctly labeled and formatted. Incorrect labels or data format might lead to issues during the validation process, including the generation of the confusion matrix.

3. **Software Version**: Make sure you are using the latest version of the YOLOv8 repository. If there was a bug in the version you are using, it might have been fixed in a more recent update.

4. **Graph Generation**: The confusion matrix generation might depend on additional libraries or settings within the YOLOv8 framework. Verify that all dependencies are installed and that the necessary graph generation code is being executed.

5. **Code Customization**: If you have made any custom changes to the YOLOv8 codebase, ensure that these changes have not inadvertently disabled or altered the confusion matrix generation process.

6. **File Paths**: The links you provided for the confusion matrix images seem to be incorrect, as they point to a non-existent location on GitHub. Ensure that the file paths for saving the confusion matrix images are correct and that the code has write permissions to those paths.

7. **Error Handling**: Check the console output or log files for any errors or warnings that might indicate why the confusion matrix is not being generated. There might be clues in the logs that can help you troubleshoot the issue.

To resolve the issue, you can start by checking the above points. If the problem persists, you may want to reach out to the Ultralytics team or the community on GitHub issues or discussions for further assistance. Provide as much detail as possible, including the exact commands you are running, your environment setup, and any error messages you are receiving.

**Comment by FahadRamzan** [comment]
I am unable to figure out the issue can you help me out

**Comment by FahadRamzan** [comment]
/content
Ultralytics YOLOv8.1.14 🚀 Python-3.10.12 torch-2.1.0+cu121 CUDA:0 (Tesla T4, 15102MiB)
engine/trainer: task=obb, mode=train, model=yolov8s-obb.pt, data=/content/datasets/GWHD2021-3/data.yaml, epochs=30, time=None, patience=50, batch=8, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train2, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None, tracker=botsort.yaml, save_dir=runs/obb/train2
Downloading https://ultralytics.com/assets/Arial.ttf to '/root/.config/Ultralytics/Arial.ttf'...
100% 755k/755k [00:00<00:00, 25.9MB/s]
2024-02-16 20:15:00.171958: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-02-16 20:15:00.172018: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-02-16 20:15:00.173381: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
Overriding model.yaml nc=15 with nc=1

                   from  n    params  module                                       arguments                     
  0                  -1  1       928  ultralytics.nn.modules.conv.Conv             [3, 32, 3, 2]                 
  1                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]                
  2                  -1  1     29056  ultralytics.nn.modules.block.C2f             [64, 64, 1, True]             
  3                  -1  1     73984  ultralytics.nn.modules.conv.Conv             [64, 128, 3, 2]               
  4                  -1  2    197632  ultralytics.nn.modules.block.C2f             [128, 128, 2, True]           
  5                  -1  1    295424  ultralytics.nn.modules.conv.Conv             [128, 256, 3, 2]              
  6                  -1  2    788480  ultralytics.nn.modules.block.C2f             [256, 256, 2, True]           
  7                  -1  1   1180672  ultralytics.nn.modules.conv.Conv             [256, 512, 3, 2]              
  8                  -1  1   1838080  ultralytics.nn.modules.block.C2f             [512, 512, 1, True]           
  9                  -1  1    656896  ultralytics.nn.modules.block.SPPF            [512, 512, 5]                 
 10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 11             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 12                  -1  1    591360  ultralytics.nn.modules.block.C2f             [768, 256, 1]                 
 13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 14             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 15                  -1  1    148224  ultralytics.nn.modules.block.C2f             [384, 128, 1]                 
 16                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
 17            [-1, 12]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 18                  -1  1    493056  ultralytics.nn.modules.block.C2f             [384, 256, 1]                 
 19                  -1  1    590336  ultralytics.nn.modules.conv.Conv             [256, 256, 3, 2]              
 20             [-1, 9]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 21                  -1  1   1969152  ultralytics.nn.modules.block.C2f             [768, 512, 1]                 
 22        [15, 18, 21]  1   2402614  ultralytics.nn.modules.head.OBB              [1, 1, [128, 256, 512]]       
YOLOv8s-obb summary: 250 layers, 11422166 parameters, 11422150 gradients, 29.6 GFLOPs

Transferred 391/397 items from pretrained weights
TensorBoard: Start with 'tensorboard --logdir runs/obb/train2', view at http://localhost:6006/
Freezing layer 'model.22.dfl.conv.weight'
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
Downloading https://github.com/ultralytics/assets/releases/download/v8.1.0/yolov8n.pt to 'yolov8n.pt'...
100% 6.23M/6.23M [00:00<00:00, 106MB/s]
AMP: checks passed ✅
train: Scanning /content/datasets/GWHD2021-3/train/labels... 4560 images, 85 backgrounds, 0 corrupt: 100% 4560/4560 [00:05<00:00, 900.30it/s]
train: WARNING ⚠️ /content/datasets/GWHD2021-3/train/images/502d5e4a8b89aeffaa6adf13e1215d27c2951230ab06a4fcb93732117196f124_png_jpg.rf.32e207b54392766e0456b93737292e7e.jpg: 1 duplicate labels removed
train: WARNING ⚠️ /content/datasets/GWHD2021-3/train/images/b5693f35d46830f4f1b35e09f7933f76f234cc641cc8b58dfec7da533231b4ad_png_jpg.rf.e6328c220111e169ebb7d499ed2894c5.jpg: 1 duplicate labels removed
train: WARNING ⚠️ /content/datasets/GWHD2021-3/train/images/ed17cba877ba4983b072ead24480913fb4cff9f1fc0f920e07021ca4f4f2ace2_png_jpg.rf.7b8fdc4a18ddea8b3096b9895838d4ce.jpg: 1 duplicate labels removed
train: New cache created: /content/datasets/GWHD2021-3/train/labels.cache
albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01), CLAHE(p=0.01, clip_limit=(1, 4.0), tile_grid_size=(8, 8))
val: Scanning /content/datasets/GWHD2021-3/valid/labels... 1303 images, 29 backgrounds, 0 corrupt: 100% 1303/1303 [00:03<00:00, 340.40it/s]
val: WARNING ⚠️ /content/datasets/GWHD2021-3/valid/images/9f697765bd41325baaf635d5389020270c4536200f97884928c014ac1d39ce12_png_jpg.rf.26a5271f93fd5e0effb338e239fd5b45.jpg: 1 duplicate labels removed
val: New cache created: /content/datasets/GWHD2021-3/valid/labels.cache
Plotting labels to runs/obb/train2/labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.002, momentum=0.9) with parameter groups 63 weight(decay=0.0), 73 weight(decay=0.0005), 72 bias(decay=0.0)
TensorBoard: model graph visualization added ✅
Image sizes 640 train, 640 val
Using 2 dataloader workers
Logging results to runs/obb/train2
Starting training for 30 epochs...

**Comment by glenn-jocher** [comment]
@FahadRamzan the confusion matrix is a valuable tool for understanding the performance of a classification model, including object detection models like YOLOv8 OBB. It provides insights into the types of errors the model is making. However, if the confusion matrix is not being generated as expected, there could be several reasons for this issue.

Here are some steps you can take to troubleshoot and resolve the problem:

1. **Check Version Compatibility**: Ensure that you are using the latest version of the Ultralytics YOLOv8 repository, as updates and bug fixes are regularly made.

2. **Review Validation Settings**: Make sure that the validation settings are correctly configured to generate confusion matrices. This includes setting the `plots` argument to `True` during validation.

3. **Inspect Data Labels**: Verify that the validation dataset is correctly labeled and formatted as per YOLOv8's requirements. Incorrect labels or formats can lead to issues during the generation of confusion matrices.

4. **Examine Log Output**: Look at the console output or log files during validation for any error messages or warnings that might indicate why the confusion matrix is not being generated.

5. **Check File Paths**: The links you provided for the confusion matrix images seem to be incorrect. Ensure that the file paths for saving the confusion matrix images are correct and that the directory has write permissions.

6. **Review Code Changes**: If you have made any modifications to the YOLOv8 codebase, review those changes to ensure they are not causing the issue.

7. **Run with Default Configuration**: Try running the validation with a default configuration to see if the confusion matrix is generated. If it works with the default settings, you can incrementally introduce your custom settings to identify the cause.

8. **Check for Updates or Issues**: Look for any recent updates in the Ultralytics YOLOv8 repository that might address this issue. Also, check the issues section to see if others have reported similar problems and if there are any suggested solutions.

If after trying these steps the confusion matrix is still not being generated, you can reach out to the Ultralytics team with detailed information about your setup, dataset, and the steps you've taken. This will help them provide more targeted assistance.

Remember to replace the placeholders in the image URLs with the correct paths where your confusion matrix images should be saved. If the images are not being saved, it could be due to incorrect paths or permission issues.

**Comment by Burhan-Q** [comment]
I am able to reproduce this issue with the `dota8` dataset

```py
from ultralytics import YOLO

model = YOLO("yolov8n-obb.pt")
results = model.val("dota8.yaml")

result.confusion_matrix.tp_fp()
>>> (array([ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]),  
     array([ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]))
```

Note, this is a TODO noted here https://github.com/ultralytics/ultralytics/blob/8aa73c447022d022016bf7c7e0084786bf554111/ultralytics/models/yolo/detect/val.py#L135

**Comment by Petros626** [comment]
@Burhan-Q still the case, that OBB does not support confusion_matrix?

**Comment by Burhan-Q** [comment]
> @Burhan-Q still the case, that OBB does not support confusion_matrix?

This feature was added in #8299

**Comment by Petros626** [comment]
@Burhan-Q thanks for the clarification. 
To produce the confusion matrix during validation this is the call `result.confusion_matrix.tp_fp()`? Can the matrix also be created during training with param `val=True`?

**Comment by Burhan-Q** [comment]
@Petros626 if you just want the confusion matrix calculated during validation, you'd only need to use `result.confusion_matrix.matrix` but if you want to calculate it yourself, you can use the `tp_fp` property instead. For training, by default you'll get the final validation confusion matrix with

```python
from ultralytics import YOLO

model = YOLO("yolo11.pt")
result = model.train(data="coco8.yaml")

result.confusion_matrix.matrix
```

but if you want to save the confusion matrix raw data for each epoch, you'll need to modify the source code manually. Remember, there is a confusion matrix plot that is saved from both the validation and training output, so if will you need is the plot, check the output directory.

**Comment by Petros626** [comment]
@Burhan-Q thank you for the insights, it's clear now.

Is there a general difference if I validate during training or afterwards? I can't decide, which would be better.

**Comment by Burhan-Q** [comment]
The final validation is always run when training to show the performance of the custom trained model. The `val=False` argument for training just means that the validation won't run for _every_ epoch, just the last one.

**Comment by Petros626** [comment]
maybe an update in the documentation like this would help to better understand.

"Enables periodic validation during training, allowing for evaluation of the model's performance on a separate validation dataset after each epoch. If val=False, validation is skipped during intermediate epochs and only performed once at the end of training to assess the final model performance."

**Comment by Petros626** [comment]
@Burhan-Q how can I change the visualized graphs (using `wandb`)  with specific metrics during training job to a specific metric (e.g. `ap50`) of `metrics.py`?

**Comment by glenn-jocher** [comment]
Thank you for the documentation suggestion! This clarification would indeed help users understand the validation behavior better. We welcome community contributions through our [Contributing Guide](https://docs.ultralytics.com/help/contributing) if you'd like to submit this improvement via Pull Request. For OBB-specific metrics implementation details, refer to the [OBB Validation documentation](https://docs.ultralytics.com/tasks/obb/#val).
